DPG/Tasks/AOTTrack/tagAndProbeDmesons.cxx

```diff
@@ -233,18 +262,103 @@ struct TagTwoProngDisplacedVertices {
+  /// \return 1 if the TagParticles belong to the chosen D meson
+  ///                 AND the D meson has the correct probe as one of its daughters;
+  ///                 0 al the other cases.
+  template <typename P, typename T>
+  int getTagOrigin(T const& firsTrack,
+                   T const& secondTrack,
+                   P const& particlesMc,
+                   const uint8_t channel,
+                   std::vector<int>& pdgDecayMothers)
+  {
+    auto pdgTagMother = 0;
+    int pdgProbeParticle = -1;
+
+    if (channel == aod::tagandprobe::TagChannels::DplusToKPiPi) {
+      pdgTagMother = constants::physics::Pdg::kDPlus;
+      pdgProbeParticle = 321; // Ka
+    } else if (channel == aod::tagandprobe::TagChannels::DsOrDplusToKKPi) {
+      pdgTagMother = constants::physics::Pdg::kPhi;
+      pdgProbeParticle = 211; // Pi
+    } else {
+      pdgTagMother = constants::physics::Pdg::kD0;
+      pdgProbeParticle = 211; // Pi
+    }
+
+    if (!firsTrack.has_mcParticle() || !secondTrack.has_mcParticle()) {
+      return -999;
+    } else {
+      auto firstMcTrack = firsTrack.template mcParticle_as<aod::McParticles>();
+      auto secondMcTrack = secondTrack.template mcParticle_as<aod::McParticles>();
+      auto firstTrackMotherId = RecoDecay::getMother(particlesMc, firstMcTrack, pdgTagMother, true);
+      auto secondTrackMotherId = RecoDecay::getMother(particlesMc, secondMcTrack, pdgTagMother, true);
+
+      if ((firstTrackMotherId == secondTrackMotherId) && (firstTrackMotherId != -1)) {
+        auto particleMother = particlesMc.rawIteratorAt(firstTrackMotherId);
+
+        /// π±π± for D± → K∓π±π± decays
+        if (channel == aod::tagandprobe::TagChannels::DplusToKPiPi) {
+          auto particleMother = particlesMc.rawIteratorAt(firstTrackMotherId);
+          auto daughters = particleMother.template daughters_as<aod::McParticles>();
+          // Check if the probe is within the mother's particle daughters
+          for (auto& daughter : daughters) {
+            if (std::abs(daughter.pdgCode()) == pdgProbeParticle) {
+              return 1;
+            }
+          }
+        }
+
+        ///  K∓K± for φ from Ds± or D± → φπ± decays
+        ///  K∓π± for D0 from D±* → D0π± decays
+        else {
+          for (auto pdgGrandMother : pdgDecayMothers) {
+            auto grandMotherId = RecoDecay::getMother(particlesMc, particleMother, pdgGrandMother, true);
+            auto particleGrandMother = particlesMc.rawIteratorAt(grandMotherId);
+            if (grandMotherId != -1) {
+              auto daughters = particleGrandMother.template daughters_as<aod::McParticles>();
+              // Check if the probe is within the GrandMother's particle daughters
+              for (auto& daughter : daughters) {
+                if (std::abs(daughter.pdgCode()) == pdgProbeParticle) {
+                  return 1;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    return 0;
```

Comment:
here you return `0`, while above when either one of the daughter does not have a MC particle you return `-999`. Is it what you want? do you really want to separate the two cases with the return? Nothing against it, I just raise the point in case you did not ask it yourself.

Reply:
You are right. We don't need this kind of further distinguish. Thank you!